nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/handler/SMTPMessageHandlerFactory.java

```diff
@@ -0,0 +1,147 @@
+
+            StopWatch watch = new StopWatch(false);
+
+            SMTPServer server = messageContext.getSMTPServer();
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            byte [] buffer = new byte[1024];
+            int rd;
+
+            while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1) {
+                baos.write(buffer, 0, rd);
+            }
+            if (baos.getBufferLength() > server.getMaxMessageSize()) {
+                throw new TooMuchDataException("Data exceeds the amount allowed.");
+            }
+
+            baos.flush();
+            this.messageBody = baos.toByteArray();
+
+
+            X509Certificate[] certificates = new X509Certificate[]{};
+
+            String remoteIP = messageContext.getRemoteAddress().toString();
+            String helo = messageContext.getHelo();
+            if (messageContext.getTlsPeerCertificates() != null ){
+                certificates = (X509Certificate[]) messageContext.getTlsPeerCertificates().clone();
+            }
+
+            SmtpEvent message = new SmtpEvent(remoteIP, helo, from, recipient, certificates, messageBody);
+            try {
+                    // Try to queue the message back to the NiFi session
+                    incomingMessages.put(message);
+            } catch (InterruptedException e) {
+                // Throws an error to the  incoming server alerting about issues queuing the message in NiFi
+                logger.error("Hit an error sending message back to NiFi main thread {}. Sending and error back to SMTP client and discarding message", new Object [] {e});
+                throw new DropConnectionException(451, "NiFi something went wrong while processing you message, please retry again later");
+            }
+
+            // Once message has been sent to the queue, it should be processed by NiFi onTrigger,
+            // a flowfile created and its processed status updated before an acknowledgment is
+            // given back to the SMTP client
+            try {
+                synchronized(message) {
+                    while(! message.getProcessed()) {
+                        // Check to see if it is too late...
+                        final long serverTimeout = TimeUnit.NANOSECONDS.convert(messageContext.getSMTPServer().getConnectionTimeout(), TimeUnit.MILLISECONDS);
+                        if ( watch.getElapsed(TimeUnit.MILLISECONDS) <=  serverTimeout) {
+                            // Will should allow for a couple of retries;
+                            message.wait(messageContext.getSMTPServer().getConnectionTimeout() / 20);
+                        } else {
+                            logger.error("Did not receive the onTrigger reponse within the acceptable timeframes");
+                            throw new DropConnectionException(451, "The processing of your message timed-out, we may have received it but you better off sending it again");
```

Comment:
@JPercivall 

Regarding the assumptions about timing. I got that. Brilliant approach that will keep in mind in the future (I believe they should all have been removed by now?)

Regarding the third race condition, isn't this condition similar to what happens in ConsumeMqtt ?

I ask because when I read [this](https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java#L294) I interpret(perhaps incorrectly?) that final test tries to remove the message from the queue (after doing its work over a peek'ed variable). If the remove fails it logs a warning.

Given Line 129 is the only remove / take / poll present I assume (also perhaps incorrectly?) that for that remove to fail, something would have gone totally belly up, but more likely than not, another thread would have won the race and consumed the object from processor queue faster than the thread logging the warning?

It this is the case, wouldn't that mean that mqttMessage also exists on both threads at the same time, having been removed from one, while the other still processing it?

Having said that I am fine to have a look at the synchronized again if you still find it the best approach (I take your judgement on that).